### PR TITLE
KAFKA-14358; Disallow creation of cluster metadata partition

### PR DESCRIPTION
**Pull request by jsancio**
With KRaft the cluster metadata topic (__cluster_metadata) has a different implementation compared to regular topic. The user should not be allowed to create this topic. This can cause issues if the metadata log dir is the same as one of the log dirs.

This change returns an authorization error if the user tries to create the cluster metadata topic.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/main/scala/kafka/server/KafkaApis.scala
Comment (mumrah) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: Can we just filter out the metadata topic once up at L1927? 

Also, should we log a warning if we find it (and remove it)?

File: core/src/main/scala/kafka/server/ControllerApis.scala
Comment (mumrah) [inline core/src/main/scala/kafka/server/ControllerApis.scala]: Similar question to below -- should we log a warning if the user passed in `__cluster_metadata`?

**Review by mumrah — COMMENTED** [review]
Thanks for the patch @jsancio. Just a couple of comments.

I tried creating `__cluster_metadata` on trunk and got a non-fatal error:

```
[2022-11-29 13:11:47,725] ERROR Encountered metadata application fault: Error applying topics delta in MetadataDelta up to 116 (org.apache.kafka.server.fault.LoggingFaultHandler)
org.apache.kafka.common.errors.InconsistentTopicIdException: Tried to assign topic ID Some(wlfZq-nVSIashndGhJaHgA) to log for topic partition __cluster_metadata-0,but log already contained topic ID AAAAAAAAAAAAAAAAAAAAAQ
```

I was able to create a different topic after this error, and the broker/controller seemed to restart fine (though the error happens again during replay). It _seems_ like a user creating this topic is not disastrous, but we should certainly avoid it :)

Edit: Apparently the topic does appear in the topic describe output, but produce and consume do not work.

File: core/src/main/scala/kafka/server/ControllerApis.scala
Comment (hachikuji) [inline core/src/main/scala/kafka/server/ControllerApis.scala:414]: Why add this?

Comment (jsancio) [reply core/src/main/scala/kafka/server/ControllerApis.scala:414]: @hachikuji Mainly because the ZK implementation returns this error message. For the test that I added in this PR, I want it to succeed in both mode (ZK and KRaft). For this, I need to either add this error message to the KRaft impl or remove this message from the ZK impl. This seems like the more backward compatible change.

**Comment by jsancio** [comment]
> Thanks for the patch @jsancio. Just a couple of comments.
> 
> I tried creating `__cluster_metadata` on trunk and got a non-fatal error:
> 
> ```
> [2022-11-29 13:11:47,725] ERROR Encountered metadata application fault: Error applying topics delta in MetadataDelta up to 116 (org.apache.kafka.server.fault.LoggingFaultHandler)
> org.apache.kafka.common.errors.InconsistentTopicIdException: Tried to assign topic ID Some(wlfZq-nVSIashndGhJaHgA) to log for topic partition __cluster_metadata-0,but log already contained topic ID AAAAAAAAAAAAAAAAAAAAAQ
> ```
> 
> I was able to create a different topic after this error, and the broker/controller seemed to restart fine (though the error happens again during replay). It _seems_ like a user creating this topic is not disastrous, but we should certainly avoid it :)
> 
> Edit: Apparently the topic does appear in the topic describe output, but produce and consume do not work.

Yes. Before this change, if the user attempts to create this topic against a KRaft cluster the CreateTopics RPC will succeed but the replica manager wont find the log in the log manager.

The behavior under the ZK implementation is also unexpected. The topic creation will succeed and the user will be able to produce and consume from it but if the brokers restart then the user will not able to produce and consume from it. This is because the code for loading log in the log manager is different from the code that creates logs in the log manager.

File: core/src/main/scala/kafka/server/KafkaApis.scala
Comment (jsancio) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: > Also, should we log a warning if we find it (and remove it)?
I don't think it should be warning. The code is behaving as expected. We can log INFO, DEBUG or TRACE.

The user is trying to create a topic that the cluster doesn't allow. We can either return a better `errorMessage` in the response or log an info message. I'll change the code log an info message.

File: core/src/main/scala/kafka/server/ControllerApis.scala
Comment (hachikuji) [reply core/src/main/scala/kafka/server/ControllerApis.scala:414]: Ok. I thought we left the error message empty whenever the message did not have additional detail over the default message in `Errors`. I think it would be fine to remove the message in both cases as well. I'll leave it up to you.

Comment (jsancio) [reply core/src/main/scala/kafka/server/ControllerApis.scala:414]: Thanks for the background information. I found another issue while trying to implement this. The default for `CreateTopicsResponse.errorMessage` is the empty string, `""`, not `null`. This means that `ApiError.messageWithFallback` doesn't behave as expected.

I am going to create an issue for this and tackle it in another PR.

**Review by mumrah — APPROVED** [review]
LGTM (pending passing tests)

**Comment by jsancio** [comment]
Thanks @mumrah . The test failures are unrelated.

**Pull request merged**
